### PR TITLE
Add fault injection for "Cannot allocate thread"

### PR DESCRIPTION
**Pull request by tavplubix**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit c53b20a77070841289c018c66ae806cc74db832e with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62266/c53b20a77070841289c018c66ae806cc74db832e/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: tests/01947_multiple_pipe_read_sample_data_ZbApel.tsv
Comment (antonio2368) [inline tests/01947_multiple_pipe_read_sample_data_ZbApel.tsv:1]: ?

**Comment by tavplubix** [comment]
Hmm
```
111547:2024.04.05 00:50:01.549954 [ 188123 ] {} <Fatal> BaseDaemon: ########## Short fault info ############
111548:2024.04.05 00:50:01.549992 [ 188123 ] {} <Fatal> BaseDaemon: (version 24.4.1.1, build id: B40A7C2DF117BA0395BB90506B2932F039452434, git hash: 9d83b61d13b13ce8ae3a35dffb8091f23bbe5940) (from thread 178260) Received signal 11
111549:2024.04.05 00:50:01.550018 [ 188123 ] {} <Fatal> BaseDaemon: Signal description: Segmentation fault
111550:2024.04.05 00:50:01.550026 [ 188123 ] {} <Fatal> BaseDaemon: Address: 0x18. Access: read. Address not mapped to object.
111552:2024.04.05 00:50:01.550033 [ 188123 ] {} <Fatal> BaseDaemon: Stack trace: 0x0000000007626548 0x0000000007843eff
111553:2024.04.05 00:50:01.550056 [ 188123 ] {} <Fatal> BaseDaemon: ########################################
111554:2024.04.05 00:50:01.550084 [ 188123 ] {} <Fatal> BaseDaemon: (version 24.4.1.1, build id: B40A7C2DF117BA0395BB90506B2932F039452434, git hash: 9d83b61d13b13ce8ae3a35dffb8091f23bbe5940) (from thread 178260) (query_id: 738be0f1-b135-4ab5-9aaa-973d066198d6) (query: DROP TABLE distributed_01099_b;) Received signal Segmentation fault (11)
111556:2024.04.05 00:50:01.550101 [ 188123 ] {} <Fatal> BaseDaemon: Address: 0x18. Access: read. Address not mapped to object.
111557:2024.04.05 00:50:01.550128 [ 188123 ] {} <Fatal> BaseDaemon: Stack trace: 0x0000000007626548 0x0000000007843eff
111674:2024.04.05 00:50:01.562092 [ 188123 ] {} <Fatal> BaseDaemon: 0.0. inlined from /fasttest-workspace/build/./src/Common/StackTrace.cpp:349: StackTrace::tryCapture()
111675:2024.04.05 00:50:01.562134 [ 188123 ] {} <Fatal> BaseDaemon: 0. /fasttest-workspace/build/./src/Common/StackTrace.cpp:318: StackTrace::StackTrace(ucontext const&) @ 0x0000000007626548
112291:2024.04.05 00:50:01.598695 [ 188123 ] {} <Fatal> BaseDaemon: 1. /fasttest-workspace/build/./src/Daemon/BaseDaemon.cpp:0: signalHandler(int, siginfo_t*, void*) @ 0x0000000007843eff
112302:2024.04.05 00:50:01.600061 [ 188123 ] {} <Fatal> BaseDaemon: Integrity check of the executable skipped because the reference checksum could not be read.
113125:2024.04.05 00:50:01.642791 [ 188123 ] {} <Fatal> BaseDaemon: This ClickHouse version is not official and should be upgraded to the official build.
113129:2024.04.05 00:50:01.642877 [ 188123 ] {} <Fatal> BaseDaemon: Changed settings: connect_timeout_with_failover_ms = 2000, connect_timeout_with_failover_secure_ms = 3000, idle_connection_timeout = 36000, s3_check_objects_after_upload = true, stream_like_engine_allow_direct_select = true, replication_wait_for_inactive_replica_timeout = 30, allow_nonconst_timezone_arguments = true, allow_experimental_parallel_reading_from_replicas = 0, parallel_distributed_insert_select = 1, log_queries = true, insert_quorum_timeout = 60000, fsync_metadata = false, http_send_timeout = 60., http_receive_timeout = 60., opentelemetry_start_trace_probability = 0.10000000149011612, max_untracked_memory = 1048576, memory_profiler_step = 1048576, log_comment = '01099_parallel_distributed_insert_select.sql', send_logs_level = 'fatal', prefer_localhost_replica = true, allow_introspection_functions = true, database_atomic_wait_for_drop_and_detach_synchronously = true, distributed_ddl_entry_format_version = 6, async_insert_busy_timeout_max_ms = 5000, load_marks_asynchronously = true, allow_prefetched_read_pool_for_remote_filesystem = false, allow_prefetched_read_pool_for_local_filesystem = false, filesystem_prefetch_max_memory_usage = 1073741824, insert_keeper_max_retries = 100, insert_keeper_retry_initial_backoff_ms = 1, insert_keeper_retry_max_backoff_ms = 1, insert_keeper_fault_injection_probability = 0.009999999776482582
170769:2024.04.05 00:50:05.544587 [ 180239 ] {16f56758-4a9a-4ee7-9085-8a7631c86932} <Trace> CreatingSetsTransform: Creating set, key: __set_18151962534936616947_148375527600104661142024.04.05 00:50:05.981009 [ 177355 ] {} <Fatal> Application: Child process was terminated by signal 11.

```

**Review by antonio2368 — APPROVED** [review]
LGTM, not sure if the failures are related

**Comment by tavplubix** [comment]
Backups:
```
==3266==ERROR: AddressSanitizer: stack-use-after-return on address 0x7f10000179a0 at pc 0x563bf5706914 bp 0x7f104c3b2a50 sp 0x7f104c3b2a48
READ of size 8 at 0x7f10000179a0 thread T760 (BackupWorker)
    #0 0x563bf5706913 in std::__1::shared_ptr<DB::IBackup const>::get[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:801:16
    #1 0x563bf5706913 in std::__1::shared_ptr<DB::IBackup const>::operator bool[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:833:16
    #2 0x563bf5706913 in DB::buildFileInfoForBackupEntry(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::IBackupEntry const> const&, std::__1::shared_ptr<DB::IBackup const> const&, DB::ReadSettings const&, std::__1::shared_ptr<Poco::Logger>) build_docker/./src/Backups/BackupFileInfo.cpp:195:9
    #3 0x563bf570a0fb in DB::buildFileInfosForBackupEntries(std::__1::vector<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IBackupEntry const>>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IBackupEntry const>>>> const&, std::__1::shared_ptr<DB::IBackup const> const&, DB::ReadSettings const&, ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>&, std::__1::shared_ptr<DB::QueryStatus>)::$_0::operator()() const build_docker/./src/Backups/BackupFileInfo.cpp:258:28
    #4 0x563bf570a0fb in decltype(std::declval<DB::buildFileInfosForBackupEntries(std::__1::vector<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IBackupEntry const>>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IBackupEntry const>>>> const&, std::__1::shared_ptr<DB::IBackup const> const&, DB::ReadSettings const&, ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>&, std::__1::shared_ptr<DB::QueryStatus>)::$_0&>()()) std::__1::__invoke[abi:v15000]<DB::buildFileInfosForBackupEntries(std::__1::vector<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IBackupEntry const>>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IBackupEntry const>>>> const&, std::__1::shared_ptr<DB::IBackup const> const&, DB::ReadSettings const&, ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>&, std::__1::shared_ptr<DB::QueryStatus>)::$_0&>(DB::buildFileInfosForBackupEntries(std::__1::vector<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IBackupEntry const>>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IBackupEntry const>>>> const&, std::__1::shared_ptr<DB::IBackup const> const&, DB::ReadSettings const&, ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>&, std::__1::shared_ptr<DB::QueryStatus>)::$_0&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23
    #5 0x563bf570a0fb in void std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::buildFileInfosForBackupEntries(std::__1::vector<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IBackupEntry const>>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IBackupEntry const>>>> const&, std::__1::shared_ptr<DB::IBackup const> const&, DB::ReadSettings const&, ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>&, std::__1::shared_ptr<DB::QueryStatus>)::$_0&>(DB::buildFileInfosForBackupEntries(std::__1::vector<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IBackupEntry const>>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IBackupEntry const>>>> const&, std::__1::shared_ptr<DB::IBackup const> const&, DB::ReadSettings const&, ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>&, std::__1::shared_ptr<DB::QueryStatus>)::$_0&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9
    #6 0x563bf570a0fb in std::__1::__function::__default_alloc_func<DB::buildFileInfosForBackupEntries(std::__1::vector<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IBackupEntry const>>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IBackupEntry const>>>> const&, std::__1::shared_ptr<DB::IBackup const> const&, DB::ReadSettings const&, ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>&, std::__1::shared_ptr<DB::QueryStatus>)::$_0, void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12
    #7 0x563bf570a0fb in void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::buildFileInfosForBackupEntries(std::__1::vector<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IBackupEntry const>>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IBackupEntry const>>>> const&, std::__1::shared_ptr<DB::IBackup const> const&, DB::ReadSettings const&, ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>&, std::__1::shared_ptr<DB::QueryStatus>)::$_0, void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16
    #8 0x563be7aeba7a in std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16
    #9 0x563be7aeba7a in std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12
    #10 0x563be7aeba7a in ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::worker(std::__1::__list_iterator<ThreadFromGlobalPoolImpl<false, true>, void*>) build_docker/./src/Common/ThreadPool.cpp:458:13
    #11 0x563be7af477c in void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/./src/Common/ThreadPool.cpp:220:73
    #12 0x563be7af477c in decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()&>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23
    #13 0x563be7af477c in decltype(auto) std::__1::__apply_tuple_impl[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()&, std::__1::tuple<>&>(void&&, std::__1::tuple<>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/tuple:1789:1
    #14 0x563be7af477c in decltype(auto) std::__1::apply[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()&, std::__1::tuple<>&>(void&&, std::__1::tuple<>&) build_docker/./contrib/llvm-project/libcxx/include/tuple:1798:1
    #15 0x563be7af477c in ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()::operator()() build_docker/./src/Common/ThreadPool.h:246:13
    #16 0x563be7ae370c in std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16
    #17 0x563be7ae370c in std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12
    #18 0x563be7ae370c in ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) build_docker/./src/Common/ThreadPool.cpp:458:13
    #19 0x563be7aefdac in void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/./src/Common/ThreadPool.cpp:220:73
    #20 0x563be7aefdac in decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23
    #21 0x563be7aefdac in void std::__1::__thread_execute[abi:v15000]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(std::__1::tuple<void, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/thread:284:5
    #22 0x563be7aefdac in void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*) build_docker/./contrib/llvm-project/libcxx/include/thread:295:5
    #23 0x563bd59ccf9a in asan_thread_start(void*) crtstuff.c
    #24 0x7f14d2864ac2  (/lib/x86_64-linux-gnu/libc.so.6+0x94ac2) (BuildId: c289da5071a3399de893d2af81d6a30c62646e1e)
    #25 0x7f14d28f684f  (/lib/x86_64-linux-gnu/libc.so.6+0x12684f) (BuildId: c289da5071a3399de893d2af81d6a30c62646e1e)
```

**Comment by tavplubix** [comment]
Wow, stress tests have finally passed, let's try again just in case

**Review by antonio2368 — APPROVED** [review]

**Pull request merged**